source/common/conn_pool/conn_pool_base.cc

```diff
@@ -379,10 +407,20 @@ void ConnPoolImplBase::drainConnectionsImpl(DrainBehavior drain_behavior) {
     return;
   }
   closeIdleConnectionsForDrainingPool();
+  if (pending_streams_.empty()) {
+    // The remaining early data clients are non-idle.
+    while (!early_data_clients_.empty()) {
+      ASSERT(early_data_clients_.back()->numActiveStreams() > 0u);
+      ENVOY_LOG_EVENT(debug, "draining_non_idle_early_data_client",
+                      "draining early data client {} for cluster {}",
+                      early_data_clients_.front()->id(), host_->cluster().name());
+      transitionActiveClientState(*early_data_clients_.front(), ActiveClient::State::Draining);
+    }
+  }
 
-  // closeIdleConnections() closes all connections in ready_clients_ with no active streams,
-  // so all remaining entries in ready_clients_ are serving streams. Move them and all entries
-  // in busy_clients_ to draining.
+  // closeIdleConnectionsForDrainingPool() closes all connections in ready_clients_ with no active
```

Comment:
both ready lists, no?  maybe have helper function which you can pass early data client and ready clients to?